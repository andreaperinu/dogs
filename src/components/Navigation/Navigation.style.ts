import styled, { css } from 'styled-components';

import { StyledTheme } from '@/styles/theme';

export const NavigationStyled = styled.nav`
${({
	theme: { deepBlue, buttonColor }
}: { theme: StyledTheme }) => css`

		ul {
			display: flex;
    	flex-wrap: wrap;
			list-style-type: none;
			padding: 0;
			margin: 0;
		}
		
		li {
			color: ${buttonColor};

			&:not(:last-child) {
				margin-right: 2rem;
			}

			a {
				text-decoration: none;
			}

			h3 {
				display: inline-block;
				color: #000;
				border-bottom: 2px solid ${deepBlue};
				padding-bottom: .2rem;
				margin-top: 0;
				transition: color .5s, border-bottom .5s;

				&:hover {
					color: ${deepBlue};
					border-bottom-color: transparent;
				}
			}
		}

	`}
`
