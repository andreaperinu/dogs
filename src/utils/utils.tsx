import { sequenceT } from 'fp-ts/lib/Apply';
import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';

import { ISelectItem } from '@/components/Select/types';
import { IBreed, IBreedResponseMessage } from '@/models/models';

export const noopComponent = () => <></>

export const maybeBreedToSelectItem = (maybeBreed: O.Option<IBreed>): O.Option<ISelectItem> => pipe(
	maybeBreed,
	O.map(({ breed }) => ({ id: breed, value: breed }))
)

export const maybeBreedsToSelectItems = (maybeBreeds: O.Option<IBreed[]>): O.Option<ISelectItem[]> => pipe(
	maybeBreeds,
	O.chain(O.fromPredicate(A.isNonEmpty)),
	O.map(A.map(({ breed }) => ({ id: breed, value: breed })))
)

export const maybeSubBreedToSelectItem = (subBreed: O.Option<string>): O.Option<ISelectItem> => pipe(
	subBreed,
	O.map(breed => ({ id: breed, value: breed })),
)

export const maybeSubBreedsToSelectItems = (breed: O.Option<IBreed>): O.Option<ISelectItem[]> => pipe(
	breed,
	O.map(({ subBreeds }) => subBreeds),
	O.map(A.map(value => ({ id: value, value: value })))
)

export const maybeBreedBySelectItem = (maybeBreeds: O.Option<IBreed[]>) => (
	maybeSelectItem: O.Option<ISelectItem>,
): O.Option<IBreed> => pipe(
	sequenceT(O.option)(maybeSelectItem, maybeBreeds),
	O.chain(([{ value }, breeds]) => A.findFirst(({ breed }: IBreed) => breed === value)(breeds))
)

export const maybeSubBreedBySelectItem = (
	maybeSelectItem: O.Option<ISelectItem>
): O.Option<string> => pipe(maybeSelectItem, O.map(({ value }) => value))

export const formatBreeds = (message: IBreedResponseMessage): IBreed[] => pipe(
	Object.keys(message),
	A.map(key => ({ breed: key, subBreeds: message[key] }))
)

export const filterBreedsBySubBreeds = (maybeBreeds: O.Option<IBreed[]>): O.Option<IBreed[]> => {

	const validSubBreedFilter = ({ subBreeds }: IBreed): boolean => pipe(subBreeds, A.isNonEmpty)

	return pipe(
		maybeBreeds,
		O.map(A.filter(validSubBreedFilter)),
	)
}
