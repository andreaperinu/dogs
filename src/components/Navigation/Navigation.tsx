import React from 'react';
import { Link } from 'react-router-dom'

import { NavigationStyled } from './Navigation.style';

export const Navigation = () => (
	<NavigationStyled>
		<ul>

			<li>
				<Link to="/">
					<h3>Random by breed</h3>
				</Link>
			</li>

			<li>
				<Link to="/list">
					<h3>List by breed</h3>
				</Link>
			</li>

			<li>
				<Link to="/sub/random">
					<h3>Random by sub-breed</h3>
				</Link>
			</li>

			<li>
				<Link to="/sub/list">
					<h3>List by sub-breed</h3>
				</Link>
			</li>

		</ul>
	</NavigationStyled>
)
