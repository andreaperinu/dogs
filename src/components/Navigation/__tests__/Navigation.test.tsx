import { setup } from '@/tests/utils';

import { Navigation as component } from '../Navigation';

describe('<Navigation />', () => {
	it('should match snapshot', () => {
		const wrapper = setup({ component, props: {}, shouldMount: true, shouldHaveRouter: true, })
		expect(wrapper).toMatchSnapshot()
	})
})