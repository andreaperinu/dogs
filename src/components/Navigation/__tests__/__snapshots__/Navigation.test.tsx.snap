// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should match snapshot 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "test",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "test",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "test",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Navigation>
      <styled.nav>
        <nav
          className="sc-bdfBwQ cIrsEs"
        >
          <ul>
            <li>
              <Link
                to="/"
              >
                <LinkAnchor
                  href="/"
                  navigate={[Function]}
                >
                  <a
                    href="/"
                    onClick={[Function]}
                  >
                    <h3>
                      Random by breed
                    </h3>
                  </a>
                </LinkAnchor>
              </Link>
            </li>
            <li>
              <Link
                to="/list"
              >
                <LinkAnchor
                  href="/list"
                  navigate={[Function]}
                >
                  <a
                    href="/list"
                    onClick={[Function]}
                  >
                    <h3>
                      List by breed
                    </h3>
                  </a>
                </LinkAnchor>
              </Link>
            </li>
            <li>
              <Link
                to="/sub/random"
              >
                <LinkAnchor
                  href="/sub/random"
                  navigate={[Function]}
                >
                  <a
                    href="/sub/random"
                    onClick={[Function]}
                  >
                    <h3>
                      Random by sub-breed
                    </h3>
                  </a>
                </LinkAnchor>
              </Link>
            </li>
            <li>
              <Link
                to="/sub/list"
              >
                <LinkAnchor
                  href="/sub/list"
                  navigate={[Function]}
                >
                  <a
                    href="/sub/list"
                    onClick={[Function]}
                  >
                    <h3>
                      List by sub-breed
                    </h3>
                  </a>
                </LinkAnchor>
              </Link>
            </li>
          </ul>
        </nav>
      </styled.nav>
    </Navigation>
  </Router>
</MemoryRouter>
`;
